integrationtests/src/test/java/org/neo4j/bolt/SessionResetIT.java

```diff
@@ -0,0 +1,361 @@
+        Set<Session> runningSessions = newSetFromMap( new ConcurrentHashMap<>() );
+        AtomicBoolean stop = new AtomicBoolean();
+        List<Future<?>> futures = new ArrayList<>();
+
+        for ( int i = 0; i < STRESS_IT_THREAD_COUNT; i++ )
+        {
+            futures.add( executor.submit( () ->
+            {
+                ThreadLocalRandom random = ThreadLocalRandom.current();
+                while ( !stop.get() )
+                {
+                    runRandomQuery( autoCommit, random, runningSessions, stop );
+                }
+            } ) );
+        }
+
+        long deadline = System.currentTimeMillis() + STRESS_IT_DURATION_MS;
+        while ( !stop.get() )
+        {
+            if ( System.currentTimeMillis() > deadline )
+            {
+                stop.set( true );
+            }
+
+            resetAny( runningSessions );
+
+            MILLISECONDS.sleep( 30 );
+        }
+
+        awaitAll( futures );
+        assertNoActiveTransactions();
+    }
+
+    private void runRandomQuery( boolean autoCommit, Random random, Set<Session> runningSessions, AtomicBoolean stop )
+    {
+        try
+        {
+            Session session = driver.session();
+            runningSessions.add( session );
+            try
+            {
+                String query = STRESS_IT_QUERIES[random.nextInt( STRESS_IT_QUERIES.length - 1 )];
+                runQuery( session, query, autoCommit );
+            }
+            finally
+            {
+                runningSessions.remove( session );
+                session.close();
+            }
+        }
+        catch ( Throwable error )
+        {
+            Throwable cause = rootCause( error );
+            if ( !isTransactionTerminatedException( cause ) &&
+                 !(cause instanceof ServiceUnavailableException) &&
+                 !(cause instanceof ClientException) )
+            {
+                stop.set( true );
+                throw error;
+            }
+            // else it is fine to receive transient errors from the driver because
+            // sessions are being reset concurrently by the main thread
+        }
+    }
+
+    private void testQueryTermination( String query, boolean autoCommit ) throws Exception
+    {
+        Future<Void> queryResult = runQueryInDifferentThreadAndResetSession( query, autoCommit );
+
+        try
+        {
+            queryResult.get( 10, SECONDS );
+            fail( "Exception expected" );
+        }
+        catch ( Exception e )
+        {
+            assertThat( e, instanceOf( ExecutionException.class ) );
+            assertTrue( isTransactionTerminatedException( e.getCause() ) );
+        }
+
+        assertNoActiveTransactions();
+    }
+
+    private Future<Void> runQueryInDifferentThreadAndResetSession( String query, boolean autoCommit ) throws Exception
+    {
+        AtomicReference<Session> sessionRef = new AtomicReference<>();
+
+        Future<Void> queryResult = runAsync( () ->
```

Comment:
Since this is CompletableFuture, we can avoid sleep and awaits here completely and just reset the session when query result future will be completed right?